### PR TITLE
Refactor api callable module

### PR DESCRIPTION
**Pull request by tbetbetbe**

### CODE REVIEW

**Comment by saicheems** [comment]
"Refactoring use decorator style" should be "Refactor..."

File: google/gax/api_callable.py [hunk 1]
```diff
@@ -33,81 +33,94 @@
...
-            return _retrying(max_attempts - 1, call, args, kwargs)
-
+def _add_timeout_arg(a_func, timeout):
+    """Updates a_func so that it gets called with the timeout as its final arg.
 
-def _retryable(call, max_attempts):
-    """Creates a function equivalent to call, but that retries on certain
-    exceptions.
+    This converts a callable, a_func, into another callable with an additional
+    positional arg
```
Comment (saicheems) [inline google/gax/api_callable.py]: End w/ period

File: google/gax/api_callable.py [hunk 2]
```diff
@@ -33,81 +33,94 @@
...
 
     Args:
-        call: A function.
-        max_attempts: The maximum number of times that the call should be
-            attempted; the call will always be attempted at least once.
+      a_func (callable): a callable to be updated
```
Comment (saicheems) [inline google/gax/api_callable.py]: Same here and below

File: google/gax/api_callable.py [hunk 3]
```diff
@@ -33,81 +33,94 @@
...
 
     Returns:
-        A function that will retry on exception.
+      callable: the original callable updated to the timeout arg
     """
-    return lambda *args, **kwargs: _retrying(max_attempts, call, args, kwargs)
 
+    def inner(*args, **kw):
+        """Updates args with the timeout."""
+        updated_args = args + (timeout,)
```
Comment (saicheems) [inline google/gax/api_callable.py]: Huh... is that a valid tuple? I guess it could be

File: google/gax/api_callable.py [hunk 4]
```diff
@@ -33,81 +33,94 @@
...
+
+    return inner
+
+
+def _page_streamable(a_func,
+                     request_page_token_field,
+                     response_page_token_field,
+                     resource_field,
+                     timeout):
+    """Creates a function equivalent to the input API call, but that performs
```
Comment (saicheems) [inline google/gax/api_callable.py]: First sentence supposed to be one line right? Maybe shorten it

File: google/gax/api_callable.py [hunk 5]
```diff
@@ -33,81 +33,94 @@
...
     page streaming through the iterable returned by the new function.
 
     Args:
-        call: A page-streaming gRPC call.
+        call: A page-streaming API call.
         request_page_token_field: The field of the page token in the request.
         response_page_token_field: The field of the next page token in the
             response.
         resource_field: The field to be streamed.
+        timeout: the timeout to apply to the API call
```
Comment (saicheems) [inline google/gax/api_callable.py]: Period

File: google/gax/api_callable.py [hunk 6]
```diff
@@ -201,30 +214,28 @@ def __init__(self, func, timeout=None, is_retrying=False,
...
+                self.max_attempts = defaults.max_attempts
+            if timeout is None:
+                self.timeout = defaults.timeout
+
+    def __call__(self, *args, **kwargs):
+        the_func = self.func
 
-    def __call__(self, request):
-        to_call = self.func
+        # Update thefunc with each of the applicable function decorators
```
Comment (saicheems) [inline google/gax/api_callable.py]: the_func

**Comment by saicheems** [comment]
lgtm otherwise

File: google/gax/api_callable.py (hunk 3 above)
Comment (tbetbetbe) [inline google/gax/api_callable.py]: Yes, it is way to express a tuple with a single value

File: google/gax/api_callable.py (hunk 1 above)
Comment (tbetbetbe) [inline google/gax/api_callable.py]: Done

File: google/gax/api_callable.py (hunk 2 above)
Comment (tbetbetbe) [inline google/gax/api_callable.py]: Done

File: google/gax/api_callable.py (hunk 4 above)
Comment (tbetbetbe) [inline google/gax/api_callable.py]: Done

File: google/gax/api_callable.py (hunk 5 above)
Comment (tbetbetbe) [inline google/gax/api_callable.py]: Done

**Comment by tbetbetbe** [comment]
> "Refactoring use decorator style" should be "Refactor..."

Done

**Pull request merged**